### ISSUE TITLE
pixel_replace poor correction for groups of bad pixels

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#7878]
_Issue [JP-3366](https://jira.stsci.edu/browse/JP-3366) was created on JIRA by [Kayli Glidic](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kglidic):_

After running some tests on the pixel_replace step, NRS has decided to keep it turned off by default for now (we'll update this in JP-3005). The reason behind this decision is the issues we encountered during testing, which we decided to document in this ticket separately.

 

We've attached slides that show the results of our tests on two different data sets (data from PID 1125 and 1128, both FS data) to this ticket.

 

In our testing of both data sets, we found that pixel_replace successfully identifies the missing pixels, but it doesn't do a great job at replacing all those pixels. Specifically, it seems to do a poor correction for groups of bad pixels (seen in PID 1128 data). In that same data set, we also noticed that this poor pixel replacement can potentially have a harmful impact on spec3 processing. In the final stage 3 product for PID 1128 data , the dips in the spectrum are larger with pixel_replace on.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661347#comment-661347):_

Did you try pixel_replace.algorithm=mingrad



I believe David Law added that algorithm to pixel_replace to work better with MIRI IFU data. Maybe it will help with NIRSPec data

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661413#comment-661413):_

[Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) - it looks like this algorithm does attempt to run if specified, but is not set up correctly for NIRSpec data.  Running on the 1128 input data, it throws a ValueError due to data array size mismatch:

> File ".../jwst/pixel_replace/pixel_replace.py", line 394, in mingrad
> 
> indx = np.where((~np.isfinite(indata))
> 
> ValueError: operands could not be broadcast together with shapes (44,436) (64,2048)

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661476#comment-661476):_

Can't say as I ever tested this on NIRSpec data; looks like it doesn't like something about the declared array sizes.  Regardless, if the issue is clustered groups of multiple bad pixels it won't do anything to those- it's only designed to extrapolate over a single missing value.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661489#comment-661489):_

Thanks, [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw). It looks like the problem is an assumption about the input model for the algorithm that doesn't hold for slit cutouts.  

Changing the code to read the size from the data shape instead of the model, around line 387:
   ysize, xsize = indata.shape
lets the algorithm complete on my test data.
 
The results for spec2 with mingrad are pretty close to the results with the default algorithm: the clusters of bad pixels are not corrected (as expected), and are assigned zero values in resampling/extraction.  The results are actually a little better for the spec3 combination, though: it leaves NaNs in place for pixels it can't replace, which allows the combination to ignore them.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661608#comment-661608):_

Hi [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) [Kayli Glidic](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kglidic), the initial implementation fails when the core of the trace is missing - a normalization factor was implicitly included, which when not equal to one would under-predict the replacement flux. I have an update to the algorithm which appears to fix the issue, though a second pair of eyes would be appreciated. The PR is here: <https://github.com/spacetelescope/jwst/pull/7886/>

With the bug fixed, increasing the number of adjacent columns appears to help. If it is expected that many datasets will have 3+ continuous pixels in the core of the trace flagged as DO_NOT_USE, perhaps the default value should be modified for NIRSpec data.

There is a different issue regarding data quality for this exposure - it looks as though two CRs/hot pixels/other have not been flagged on the edge of the trace, though neighboring pixels have been flagged. These two pixels appear anomalous and throw off some of the median profiles. Is this typical of these data or unexpected? I'll attach some cutouts. I also noticed that a significant fraction of the exposure is flagged as saturated. [^cutout_59_80_satcheck.pdf]  [^cutout_59_80.pdf] 

The fix also over-corrects due to what appears to be a bad pixel in the second dither. I'll attach some figures for this as well - with the trace pixels missing and the remaining pixel on the "shoulder" of the trace reading higher than expected, the normalization boosts the replacement higher than reasonable. Unclear if there is a solution to this sort of problem.
 [^jw01128012001_03102_00002_pixrep_profiles.pdf]  [^jw01128012001_03102_00002_pixrep.pdf]

**Comment by stscijgbot-jp** [comment]
_Comment by [Kayli Glidic](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kglidic) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661730#comment-661730):_

Hi Tyler, thanks for the quick update! I went ahead and re-ran both datasets with your recent update to the algorithm. I added the new results to the ticket: [^pixel_replace_problem_updated.pptx]. 

 

With the new update the data in PID 1125 look much better in the stage 2 products, it does a much better job at replacing those pixels. For PID 1128 data I tried n_adjacent_cols=3 (Default) and n_adjacent_cols=5 with the new algorithm update. I can confirm that increasing the number of adjacent columns does appear to help but I also see the over-correction in the second dither you mentioned. 

 

I’ll need to check with others to see if data like PID 1128 is typical, but overall the data looks much better with the update.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=661963#comment-661963):_

[Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) - I'm trying to track down the weird flagging in the 1128 data. Can you tell me which file and what region your cutout images are from?

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=662070#comment-662070):_

I have looked at the odd flagging patterns in the 1128 data (jw01128012001_03102_00001_nrs1, near x,y=77,34 in the slit cutout). I believe they are associated with the saturation at the peak value of the source.  The entire region near the core is flagged as 'jump', and it's likely there are some charge migration issues or other detector effects in the wings of the PSF, near the saturated area.  Running again with the charge_migration step on does not show the flagging pattern near pixel 77,34, but does mark the entire central region with NaN.

In short, no, I don't think this commissioning observation is a typical data set: the significant saturation causes several associated artifacts, and it's okay that the pixel_replace algorithm can't completely compensate for them.  The new changes look significantly better, even for this extreme case – thanks for the quick fix!

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=662073#comment-662073):_

(y) Thanks for the bug report!

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=662338#comment-662338):_

Fixed by #7886

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3366?focusedCommentId=663091#comment-663091):_

Fix looks good, this issue is resolved.

I will file a separate ticket for the the crash when mingrad is selected for non-MIRI data.
